BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+import com.karolismed.simple_blockchain.blockchain.model.transaction.TransactionOutput;
+import com.karolismed.simple_blockchain.blockchain.model.transaction.UnspentOutput;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.ThreadLocalRandom;
+import org.apache.commons.collections4.MultiValuedMap;
+import org.apache.commons.collections4.multimap.HashSetValuedHashMap;
+import org.apache.commons.lang3.tuple.Pair;
+
+import static com.karolismed.simple_blockchain.constants.GlobalConstants.BLOCK_ZERO_HASH;
+import static com.karolismed.simple_blockchain.constants.GlobalConstants.GENESIS_BLOCK_DIFFICULTY;
+import static com.karolismed.simple_blockchain.constants.GlobalConstants.REGULAR_BLOCK_DIFFICULTY;
+import static com.karolismed.simple_blockchain.constants.GlobalConstants.TRANSACTION_POOL_SIZE;
+import static com.karolismed.simple_blockchain.utils.StringHelper.log;
+import static java.util.Objects.isNull;
+
+public class BlockChainRunner {
+
+    private final BlockChainSetupService setupService;
+    private final MerkleTreeConstructor merkleTreeConstructor;
+
+    private List<User> users;
+    private List<Block> blockChain;
+    private List<Transaction> transactionPool;
+    private List<Transaction> pendingTransactions;
+    private MultiValuedMap<PublicKey, UnspentOutput> unspentOutputMap;
+    private Map<String, Integer> txIdBlockIndexMap;
+    private Map<PublicKey, User> pkUserMap;
+
+    private int totalTransactionsPooled;
+
+    public BlockChainRunner() {
+        setupService = new BlockChainSetupService();
+        merkleTreeConstructor = new MerkleTreeConstructor();
+
+        blockChain = new LinkedList<>();
+        users = new ArrayList<>();
+        transactionPool = new LinkedList<>();
+        pendingTransactions = new ArrayList<>();
+        unspentOutputMap = new HashSetValuedHashMap<>();
+        txIdBlockIndexMap = new HashMap<>();
+        pkUserMap = new HashMap<>();
+    }
+
+    public void init() throws Exception {
+        log("Setting up BlockChain...");
+        setupService.setupUsers(users, pkUserMap);
+
+        log("Mining genesis block...");
+        mineGenesisBlock();
+
+        log("Setting up miners...");
+        BlockingQueue<Pair<Block, Integer>> blockQueue = new LinkedBlockingQueue<>();
+        Semaphore threadSyncSemaphore = new Semaphore(11);
+        Miner[] miners = {
```

Comment:
The code itself is excessive, as you could have used either intStream or refactor the code in some way, where the initialization itself would take only a couple of lines, rather than 14.

It would be suggestable to either use intStream or a cycle to initialize multiple elements in an array.

Reply:
Addressed